### ISSUE TITLE
What is the plan for Python 3 support?

### ISSUE THREAD

**Issue by pspacek** [secdev/scapy#657]
Hello,

I'm trying to find out if there are any plans for "the original" scapy to support Python 3 or not.

The only references to real code I found is in https://github.com/secdev/scapy/issues/87#issuecomment-258094860 but it seems to me that the plan for scapy rewrite is not moving forward because repo https://bitbucket.org/secdev/scapy3-prototype2 has last commit from 2015-02-25.

Given that current Fedora (and so future of RHEL/CentOS/Scientific Linux) is migrating away from Python 2 (as well as all other Linux distros) this forces me to ask the question again:
What are the plans for Python 3 support in scapy?

Without some form plan, Fedora/RHEL/CentOS will have to implement the same approach as Debian (see https://github.com/secdev/scapy/issues/87#issuecomment-257431044), i.e. use scapy3k instead of scapy.

If this is the case, are there any plans to start merging features between scapy and scapy3k?

Thank you for your time and answers.

**Comment by guedou** [comment]
Python 3 compatibility will likely happen during 2017. However, it is difficult to know exactly when. @gpotter2 does an amazing job at pushing Python 3 support by submitting dedicated PRs.

The efforts currently happen on two labels:
- [coverage](https://github.com/secdev/scapy/issues?utf8=%E2%9C%93&q=label%3Acoverage): the goal is to enhance the code coverage in order to make sure that things won't break;
- [convergence](https://github.com/secdev/scapy/issues?utf8=%E2%9C%93&q=label%3Aconvergence): the goal is to rewrite parts of Scapy to make it Python 3 compatible code.

Code coverage is now close to 80% (it was ~50% back in December 2016). I believe that it is a good indicator that Python 3 compatibility can be achieved in 2017 without introducing regressions.

Feel free to contribute if you are interested !

**Comment by pspacek** [comment]
Oh, that is awesome! I'm on packaging side here so it is sufficient for me to know that something is happening and that we do not need to drag scapy3k into Fedora (assuming the port will happen soon).

Thank you very much!

**Comment by gpotter2** [comment]
@pspacek Scapy3k was created a quite long time ago. It has not received all the bug fixes of scapy and its development is way less active than this one.

There are no plan of merging both because of how old it is, and it wouldn't not make sense to fix the conflicts between them...

Merging the different parts of python 3 support is slow, as we need to keep a stable environment.

I have a roadmap of the next changes needed, and will pretty much work as quick as the review goes in, but there are quite a lot of things to fix :)

**Comment by Zopieux** [comment]
Does the scapy dev team plans for a Python 2 **and** 3 code base, or an alternative branch/project for Python 3 support maintained separately?

**Comment by guedou** [comment]
The goal is to make the code Python 2 and 3 compatible. Some PRs were already merged while some are waiting for reviews, comments and modifications.

Feel free to have a look and contribute.

**Comment by L1ghtn1ng** [comment]
@guedou @gpotter2 This is a fork and is python 3 only https://github.com/phaethon/scapy and IMO scapy should by python 3 only as on this day https://pythonclock.org/ python 2 is EoL. With the changes in the fork, I do not see why you could not get python 3 support much quicker if you want to work with @phaethon to do so?

**Comment by guedou** [comment]
Unfortunately, the creator of this fork never tried to help maintaining Scapy, and never pushed anything to us despite our calls to work on a shim layer. His behavior is quite aggressive and we don't want to work with him.

If you feel like helping us, we always welcome contributions !

**Comment by phaethon** [comment]
My short version of phaethon/scapy branch and our relationship with @guedou and co. In 2015 an issue was raised (not by me) on bitbucket repository about python3 support and the answer of the core team was "not at this time, if you need it fork the repository". This is exactly what I did. After that I hoped that there would be some cooperation between the repos, but no - I asked to include info about python3 compatibility, but the suggestion was refused. I wanted people to find out that it is possible to use scapy with python3 (as many are tied to python3 for one reason or another) and was pushing information about the fork across different channels. The core team considered it being aggressive and in a passive aggressive form themselves kept ignoring if not hiding info about the python3 fork.

I believe @guedou would more likely either reimplement from ground 0 python3 compatibility (as it was done for pcapng support after it was implemented in scapy3k) or wait for someone to bring the change from scapy3k to scapy (like it was done with Windows reimplementation, which eliminated libdnet, which was copied from scapy3k by a contributor). I would love to work in a cooperative way, but don't believe @guedou - he never invited me. Posting all scapy3k changes as one big PR would not make sense. 

I would believe it is possible to solve the situation by somebody from the community moderating the situation, but I wouldn't believe @guedou would let this happen. @guedou - fix me if I am wrong.

**Comment by phaethon** [comment]
and, btw, I wouldn't say it would be so hard to go through the differences and resync the repositories. With a trained eye it is quite quick to go through the commits. But for some people it would be too embarrassing to credit any of my effort, which made hundreds (measured by stars on github) or may be thousands of people to keep using scapy for last couple of years even if they are choosing to be with python3.

**Comment by gpotter2** [comment]
@phaethon Hello,
You're not forced by anyone to cooperate. However your help would be greatly appreciated :) as you know pretty much all we need to do to port scapy to python 3.

It does not really make sense to have two concurrent repositories. Moreover it's not useful to remind us what you did add first on yours.

In a near future, this repository will support python 3, with or without your help. As not being the official repo, it's very likely that your 2 years work will get lost, which would be quite sad.

I would be really happy to get some help with the porting :) Hope you will understand what I mean by that

**Comment by guedou** [comment]
Well, that's a convenient way to relate what really happened. It is quite distracting and fun to read. Thanks that was refreshing !

**Comment by phaethon** [comment]
@gpotter2 - a large part of the help you need is residing in a form of code in phaethon/scapy. You are welcome to ask me any questions that may arise. Fire up an issue on phaethon/scapy and I will prioritize responding to you over other issues. Actually, what I think has to be discussed first is the "interface" - how to treat strings and bytes, and what gets returned by different functions, and what expected at an input of the various function calls. scapy3k uses bytes for most things, but in some cases it may make sense to return/expect a string (e.g. domain name). It would be sad if secdev/scapy implemented incompatible interface to what scapy3k has, as I would believe many people working with python3 already have some code developed.

And, yes, I would appreciate including credit to me in a Readme of this repository as it has de facto used (I guess it still uses) parts of code copied from phaethon/scapy. This would be a nice and friendly invitation for further cooperation. Is that unreasonable?

**Comment by p-l-** [comment]
I totally support @guedou point of view here. I don't think Scapy's code uses your code from phaethon/scapy, but I might be wrong.

**Comment by L1ghtn1ng** [comment]
If everyone can put things aside I would say that pulling in the code from the fork and having credit for it where credit is due would be I think for the best of everyone? and in turn the community. I do not mind acting as a mediator between the 2 parties.

At the end of the day I want everyone to be happy and the community to benefit from the work that people have done in the fork and original.

One other thing I would like to ask is that can we get more documentation for scapy.

Sent from phone

________________________________

**Comment by p-l-** [comment]
@L1ghtn1ng I don't think we need a mediator here, but thanks for your offer. We want to have Scapy work with both Python 2 & 3 and the same code base, and we are working on it.

**Comment by irl** [comment]
I had previously been told that Python 3 support was not a priority for Scapy (https://github.com/secdev/scapy/issues/87#issuecomment-258094860) but it appears that there's been some backtracking here.

As I have @phaethon's fork packaged as python3-scapy in Debian currently, that is as far as Debian is concerned the "official" Python 3 variant of Scapy.

@phaethon Do you have a test suite that could be applied to any Python 3 porting work in this repository to ensure that if we change the upstream source for the python3-scapy package in Debian back to @secdev's repository that we wouldn't break existing uses of Scapy on Python 3 in Debian?

**Comment by p-l-** [comment]
@irl Scapy comes with an important test suite, and when it can use Python 3 we will be able to use the very same tests with both Python 2 and 3.

**Comment by irl** [comment]
@p-l- You're missing the point. I need to ensure that whatever Python 3 support for Scapy looks like, anything that worked in @phaethon's fork still works. Otherwise, I don't see a path for switching the "python3-scapy" package to be being built from the secdev sources.

**Comment by p-l-** [comment]
@irl you're missing the point. The fork you're using as a base is **not** Scapy. It's a fork, based on Scapy, that should probably have another name to avoid confusion.

Its API and behavior already differ from Scapy's API. And Scapy's API and behavior are what exist in this repository.

**Comment by irl** [comment]
Based on the previous comment I had received, linked above, I did not believe that there would be Python 3 support in Scapy and so this was the only way to include it in Debian.

The python3-scapy package provides a certain API and those that have installed the package in Debian will be upset when everything breaks because the API is different.

I only care about what is in Debian and maintaining compatibility across versions within Debian. My priorities are the users of Debian and high quality free software.

This is only an issue for the secdev developers if having their Scapy in Debian is the desirable outcome for them. Really, this is an issue for me, but it can be helped by a test suite from @phaethon so that I can come up with a series of patches to try to stop breakage from occurring due to a change in which project is considered the upstream for the package.

At the very least, I'd like to have an idea of how much breakage there would be.

If having Scapy in Debian is not something that really matters to you and you don't care, then that is also fine and I can follow up with @phaethon somewhere else.

**Comment by mcpat** [comment]
If you really care what is in Debian it would be best to transition the python3-scapy package to a different one with a different name. Because using the same name for two different (even if diverged) things is not meaningful. The debian package should have been named python3-scapy3k from the beginning because that's how it's called upstream...

**Comment by irl** [comment]
@mcpat This is not useful. I had previously been told that there would not be Python 3 support in Scapy and that @phaethon's fork was considered the Python 3 version going forward. This is the situation and this is what needs resolving.

**Comment by p-l-** [comment]
The code you are currently packaging as "python3-scapy" is **not** Scapy. **This code is Scapy**. Python 3 support or not does not change that.

**Comment by gpotter2** [comment]
@mcpat is very right.
Scapy3k is a very different fork, and it should never have been called 'scapy for python 3'.

You have been told that python 3 support was 'not a priority' 10 months ago. This isn't the case anymore. We're currently working on it.

To provide a high-quality software, we're processing the migration much slower than scapy3k did, in order to provide a backward compatibility for python 2.

Please also note that secdev's scapy has received way more bug fixes, as the official repository, than any other fork.
Of course, we all want scapy to be distributed on other plateformes, but we have not finished the migration yet, so all we can say is that we are working on it.

**Comment by irl** [comment]
This conversation appears to be a waste of everyone's time. I would appreciate it if someone could ping me when this repository's support for Python 3 is nearing stability and I'll look at possible paths for the change at that time.

**Comment by mcpat** [comment]
Hm why exactly is it not helpful? I just proposed a way to clean the current situation while not frustrating the users of the scapy3k API. I mean you decided yourself to package it under python3-scapy even though it was clear back than, that ongoing development and extensions are done in this repo here - and a merge path was and is not established!
Also, nobody said back than that support for Python3 was not envisioned. It just did not make it to the roadmap...

**Comment by p-l-** [comment]
FWIW, Scapy is now usable with Python 3. Most tests pass (we still have to fix some parts of the code), and you can report bugs with Python 3 (see #936 for example). Thanks to all the contributors, and to @gpotter2 particularly!

**Comment by johnthagen** [comment]
@p-l- Is there a plan/timeline for when a PyPI release supporting Python 3 would be uploaded? Perhaps a "Dev" or alpha release people could start trying out?

**Comment by p-l-** [comment]
@johnthagen not for now. We have to get all tests pass before that. In the meantime, users wanting to test Scapy with Python 3 are encouraged to use this repository!

**Comment by q2dg** [comment]
Hello.
If I write in shell: python2 scapy, it's all right.
If I write in shell: python3 scapy, I get this error:

Traceback (most recent call last):
  File "/usr/bin/scapy", line 23, in <module>
    from scapy.main import interact
ModuleNotFoundError: No module named 'scapy'

I've installed scapy from official repositories of Fedora 27 and its main file is located in /usr/bin
Thanks

**Comment by gpotter2** [comment]
@q2dg Please note that there has been no new release since python 3 support: you need to use the github version by downloading this repo over https://github.com/secdev/scapy/archive/master.zip
Then use
ˋpython3 setup.py install`

**Comment by q2dg** [comment]
@gpotter2 Oooh, thanks a lot! Doing like you say its works well! Sorry for the inconvenience.
(Maybe pointing this simple trick in README.md could be a good idea)

**Comment by irl** [comment]
Would you consider making even a beta or RC release? If you can do this, I'll make an upload to Debian experimental with a package to make this available for testing.

**Comment by guedou** [comment]
We will make an RC as soon as we have merge #861. It fixes some remaining issues, and enables automatic Travis tests.

**Comment by johnthagen** [comment]
#861 has landed! 🎆

**Comment by irl** [comment]
If there can be an RC release, I'm currently releasing the next version of PATHspider (irl/pathspider#188) and can test this against that. There is then only one other Debian package with a dependency on python3-scapy and then there should be a clear path to having python3-scapy switch upstream back to this repository.

**Comment by p-l-** [comment]
@irl I think @guedou planned a release for tomorrow morning.

**Comment by irl** [comment]
Hm ok, do you see any significant changes happening between now and then? I can do my PATHspider testing with a local Debian package based on the current master perhaps?

**Comment by p-l-** [comment]
I don't think so. Maybe #1024 could land in the meantime since it increases coverage and it's a tiny patch. Also, if @gpotter2 makes a PR to add Windows x86_64 tests. Anyway nothing that would impact the API.

**Comment by irl** [comment]
Cool, thanks.

**Comment by guedou** [comment]
I just tagged 2.4-rc1, see https://github.com/secdev/scapy/releases

We need to merge some more PR before v2.4.0 but this will happen in the next weeks.

**Comment by irl** [comment]
Ok, from my testing this is looking good for the requirements of PATHspider. I've not tested dhcpcanon yet though. I've opened phaethon/scapy#231 to look at making this co-installable and a proof of concept patch was working, until I messed it up and broke the patch which will need to be rewritten.

* python3-scapy will be renamed to python3-scapy3k (using package name of "scapy3k" for import)
* the python3-scapy3k will breaks+replaces python3-scapy << 0.21-1
* the src:scapy package will then have a new python3-scapy binary package built from it (using package name of "scapy" for import)
* both pathspider and dhcpcanon will then depend on scapy from this repository, keeping the fork available for those that need its API

**Comment by johnthagen** [comment]
@guedou Will you wait until v2.4.0 before making a PyPI release or were you also planning to release a "-dev" release onto PyPI?

**Comment by guedou** [comment]
I did not manage to do it today. I will try to make one tomorrow.

**Comment by guedou** [comment]
@johnthagen please try `pip install --pre scapy` that will install the latest `rc` (currently rc2).

**Comment by johnthagen** [comment]
@guedou I installed 2.4rc2 into a Python 3.6 virtual environment, played around making a couple sample packets, and everything worked! 🥇

**Comment by p-l-** [comment]
@irl I've tested [dhcpcanon](/juga0/dhcpcanon) which works fine (which is normal since it uses Scapy for Python 2 already).

**Comment by p-l-** [comment]
@guedou pressed the "release" button for 2.4.0, so I guess this can be considered fixed :)

Thanks a lot to everybody involved in this great project!

**Comment by gpotter2** [comment]
Good news: I was [blocked](https://help.github.com/articles/blocking-a-user-from-your-organization/) from phaethon/scapy, so that I can't redirect mis-leaded users to this repo instead, when they are having issues that we fixed back in 2016... Thanks @phaethon

![image](https://user-images.githubusercontent.com/10530980/39663096-0555a490-506d-11e8-8b58-c6d1bc944398.png)

Advice to all package maintainers (knowkingly @anthraxx, @irl @pspacek and others...): get rid of python3-scapy ASAP... It has a massive amount of unresponded issues, growing each day https://github.com/phaethon/scapy/issues and the guy seems to be willing to keep his github stars... https://github.com/phaethon/scapy/issues/231.

The name conflict is getting tons of issues (especially on stackoverflow) where people ask themselves why only the python 2 version is working...

Even the last arguments he had (we have CAN support), felt apart when its author moved to this repo (https://github.com/secdev/scapy/pull/1363), and we have about x8 more bug fixes than he does (just see the commits number, that used to be the same project...)

I'm tired of those confligs, and can't get why this thing is still getting published.

Have all a good day,

**Comment by Zopieux** [comment]
Open-source development is a fantastic and fulfilling initiative, but comes with its share of human vices. You can't reasonably ask for everybody to play fairly and without pride. @phaethon made a great contribution to bring Python 3 support to scapy which, at the time, may have been lagging behind – however good their reasons.

If @phaethon doesn't want to collaborate with upstream, so be it. It's still his/her right to keep the GitHub repo online. Let's stay respectful of the work accomplished by everybody and gently educate users about the issue, even if that takes time.

**Comment by phaethon** [comment]
Thanks @Zopieux for bringing a share of fairness into the discussion. 
phaethon/scapy works well for many users since 2015, and I have no short term intention to delete the repo. Contributions are welcome. If there will be any critical flaw, it will get fixed.
If there are constructive suggestions for the benefit of users I am always open to discuss them. 
My ages-old suggestion is to put a text explaining options of scapy implementations and existence of the fork in all appropriate places. 
One maintainer of this repo even went so far as deleting valid information (information explaining existence of the fork and a link to it) from Wikipedia. That guy would be probably too ashamed to admit it in the public. And I have no intention of meaningless fight over such issues. I would be happy if somebody would put a neutral text over there, on the Readme of this repo, and I am happy to make objective additions to phaethon/scapy repo's Readme.
And, yes, if someone embarrasses users of my repo (e.g. claiming that something is fixed for 2 years, while actually it is not, or not even reading the actual issue and copy pasting their text), I keep my right to block the intimidating user.

**Comment by Zopieux** [comment]
Well, thanks for keeping the discussion alive. Having a page explaining the state of the various forks is a good idea, as long as we play fair and use the same wording in upstream and forks. I guess the real issue is the branding: scapy3k is *not* scapy with 3k compatibility (because scapy now has it), it's a fork that has diverged significantly over time.

I just saw the phaethon/scapy#231 discussion about changing the module name and I appreciate the decision to implement the change. I hope the secdev people are okay with the `scapy3k` name.

@phaethon, would you be okay with a prominent line in your README stating this is a fork of the *original scapy project* with a link?
@p-l-, would you be okay with a table somewhere in the README listing the forks with links?

**Comment by gpotter2** [comment]
I'm not asking for the repo to be deleted, but removed from beeing distributed as `scapy-python3`.
@phaethon Feel free to find a issue where I reported a fix on secdev that was not real. FTR, you can find on secdev/scapy tons of PRs taken from issues reported on your repo.
E.g:
- https://github.com/secdev/scapy/issues/1135
- https://github.com/secdev/scapy/issues/737
- https://github.com/secdev/scapy/issues/960

We are fixing those and you are not.

I'm only tired of responding to people "please switch to secdev/scapy and retry", see stackoverflow: many people are asking things we fixed. It makes our work hader, and waste our time...

You know it yourself, your code has good stuff, but is incredibly not updated, not maintained and buggy.  This may be due to the fact that you are the single maintainer of it, or that you get way less contributions, since we support Python 3 too.

Your repository is currently in a state, where it attracts users because they misunderstand the name, rather than for the code it contains. You surely don't want to admit it, but that's reality.

**Comment by phaethon** [comment]
@Zopieux I am fully supportive on placing objective bilateral linking with information reflecting reality. Probably, we would need to agree on 'what reality is' first, but sure it is possible with a reasonable mediator in between the maintainers of both repos.

**Comment by gpotter2** [comment]
@irl Hello ! Please update the debian issue: scapy3k is now kamene, so I'm guessing it should be `python-kamene` rather than `python-scapy3k`

**Comment by irl** [comment]
I filed bugs about this a while ago, waiting on ftp master to process https://bugs.debian.org/898105. Can't really do anything until that happens.

**Comment by irl** [comment]
You could always email 898105@bugs.debian.org with words of encouragement.

**Comment by pacho2** [comment]
Is scapy-2.4.3 still relying on old pyx version (that is python2 only) for 2D graphs or documentation needs updating?
https://scapy.readthedocs.io/en/latest/installation.html

Thanks for the info :)

**Comment by gpotter2** [comment]
@pacho2 No, Scapy will work with any PyX version. It uses it only for optional features and will work without it. PyX >= 1.13 supports Python 3 (https://github.com/pyx-project/pyx).

I'll however lock this issue because this is sorta unrelated, and I don't want to keep pinging so many people. Please open a different issue or ask on the gitter if you need additional advices :-)
